include/circt/Dialect/HW/HWAttributes.td

```diff
@@ -11,14 +11,81 @@
+    ends in a `/` it is considered an output directory.
+
+    When creating an `OutputFileAttr` the `name` and `directory` should be
+    normalized, with uneeded indirection to the target file removed.  This is
+    so that `OutputFileAttrs` which are pointing at the same file will be
+    equivalent. The `getWithFilename` function will properly handle this
+    process.
+
+    When ExportVerilog runs, one of the files produced is a list of all other
+    files which are produced. The flag `excludeFromFileList` controls if this
+    file should be included in this list. If any `OutputFileAttr` referring to
+    the same file sets this to `true`, it will be included in the file list.
+    This option defaults to `false`.
+
+    For each file emitted by the verilog emitter, certain prelude output will
+    be included before the main content. The flag `includeReplicatedOps` can
+    be used to disable the addition of the prelude text. All `OutputFileAttr`s
+    referring to the same file must use a consistent setting for this value.
+    This option defaults to `true`.
+
+    Examples:
+    ```mlir
+      #hw.ouput_file<"/home/tester/t.sv">
+      #hw.ouput_file<"t.sv", excludeFromFileList, includeReplicatedOps>
+    ```
+  }];
+  let mnemonic = "output_file";
+  let parameters = (ins "::mlir::StringAttr":$filename,
+                        "::mlir::BoolAttr":$excludeFromFilelist,
+                        "::mlir::BoolAttr":$includeReplicatedOps);
+  let builders = [
+    AttrBuilderWithInferredContext<(ins
+                                   "::mlir::StringAttr":$filename,
+                                   "::mlir::BoolAttr":$excludeFromFileList,
+                                   "::mlir::BoolAttr":$includeReplicatedOps), [{
+      return get(filename.getContext(), filename, excludeFromFileList,
+                 includeReplicatedOps);
+    }]>,
+  ];
+
+  let extraClassDeclaration = [{
+    /// Get an OutputFileAttr from a string filename, canonicalizing the
+    /// filename.  
+    static OutputFileAttr getFromFilename(::mlir::MLIRContext *context,
+                                          const ::mlir::Twine &filename,
+                                          bool excludeFromFileList = false,
+                                          bool includeReplicatedOps = false);
+
+    /// Get an OutputFileAttr from a string filename, resolving it relative to
+    /// `directory`. If `filename` is an absolute path, the given `directory`
+    /// will not be used.
+    static OutputFileAttr getFromDirectoryAndFilename(
+                                          ::mlir::MLIRContext *context,
+                                          const ::mlir::Twine &directory,
+                                          const ::mlir::Twine &filename,
+                                          bool excludeFromFileList = false,
+                                          bool includeReplicatedOps = false);
```

Comment:
I'm not in love with this function name, but it can't have the same name as `getFromFilename` because the compiler has no problem converting `string->bool` intsead of `string->twine` and preferring the other function.

Reply:
How about dropping some default arguments?